### PR TITLE
Add tracer sampler 

### LINKED ISSUE #70: how to set client sample rate
I read the trace.go code ,I find the sample set 1。 
image

### PR DESCRIPTION
**Pull request by xbkaishui**
Fix #70 , support client side trace sampling service.

### CODE REVIEW

File: sampler.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+// IsSampled implements IsSampled() of Sampler.
+func (s *ConstSampler) IsSampled(traceID string, operation string) bool {
+	return s.decision
+}
+
+type RandomSampler struct {
+	samplingRate float64
+	rand         *rand.Rand
+	Threshold    int
```
Comment (wu-sheng) [inline sampler.go:49]: Why `T` in upper case?

File: sampler.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+func (s *RandomSampler) init() {
+	s.rand = rand.New(rand.NewSource(time.Now().Unix()))
+	s.Threshold = int(s.samplingRate * 100)
+}
+
+func NewRandomSampler(samplingRate float64) *RandomSampler {
+	if samplingRate < 0.0 || samplingRate > 1.0 {
+		return nil
```
Comment (wu-sheng) [inline sampler.go:64]: I think this is not a good practice, the user would expect `nil` when use sampling unless you have an err status.

Comment (wu-sheng) [reply sampler.go:64]: How about `<=0` and `>=1.0` use `NewConstSampler`?

File: sampler.go (hunk 1 above)
Comment (xbkaishui) [reply sampler.go:49]: This is just for test, random sampler is not easy to test the correctness

File: sampler.go (hunk 2 above)
Comment (xbkaishui) [reply sampler.go]: Ok, move the logic to WithSampler

File: sampler.go (hunk 1 above)
Comment (xbkaishui) [reply sampler.go]: changed Threshold to threshold, sorry I am newer to golang

File: segment.go [hunk 3]
```diff
@@ -237,7 +237,11 @@ func newSegmentRoot(segmentSpan *segmentSpanImpl) *rootSegmentSpan {
 				break
 			}
 		}
-		s.tracer.reporter.Send(append(s.segment, s))
+		//check need send segment with sample service
+		sampled := s.tracer.sampler.IsSampled(s.TraceID, s.FirstSpan.GetOperationName())
```
Comment (arugal) [inline segment.go:241]: 1. Should be sampled when the first `Span` is created, and `NoopSpan` is returned when ignored.
1. Sampling should be forced when downstream services are sampled.

Hi @xbkaishui, that's my suggestion, what do you think?

Comment (xbkaishui) [reply segment.go:241]: Hi @arugal, thanks for your cr, I have a couple of questions about what your mention
for the first one, how to decide which one is the first span ? 
for the second one, how to check when downstream services are sampled, if it is have sw8, I should consider it is force sampled?

**Comment by xbkaishui** [comment]
@arugal, I have changed the logic , please help to check if it is acceptable, thanks

File: segment.go (hunk 3 above)
Comment (arugal) [reply segment.go]: > how to decide which one is the first span ?

https://github.com/SkyAPM/go2sky/blob/fdae247dbb32f87c064aa59e557d814c2eb62fb5/trace.go#L116-L119

When `parentSpan` is `nil`.

> how to check when downstream services are sampled, if it is have sw8, I should consider it is force sampled?

Yes, when `sw8` header exist.

 
I think that the code should be added to `trace#CreateLocalSpan` or `trace#createNoop`.

Comment (xbkaishui) [reply segment.go]: Hi @arugal  Thanks for you help,  changed as your suggestion. please help review again

Comment (arugal) [reply segment.go]: 1. `CreateEntrySpan` and `CreateExitSpan` should judge the `CreateLocalSpan` returns `NoopSpan` or not. Specifically `CreateExitSpan`, if `NoopSpan` returns `span type is wrong` error.

2. `traceID` is not required for `IsSampled`, please recheck.

**Comment by xbkaishui** [comment]
@arugal , changed as your request, please help to check again. Thanks

File: trace.go [hunk 4]
```diff
@@ -136,6 +151,10 @@ func (t *Tracer) CreateExitSpan(ctx context.Context, operationName string, peer
 	if err != nil {
 		return nil, err
 	}
+	_, ok := interface{}(s).(NoopSpan)
+	if ok {
+		return nil, errors.New("span type is wrong")
+	}
```
Comment (arugal) [inline trace.go:157]:
```suggestion
	noopSpan, ok := s.(NoopSpan)
	if ok {
	         // Ignored, there is no need to inject SW8 in the request header
		return noopSpan, nil
	}
```

**Review by arugal — COMMENTED** [review]
Sorry for the late reply. I may not have described it clearly before.

File: trace.go (hunk 4 above)
Comment (xbkaishui) [reply trace.go]: Got it, changed

**Review by arugal — APPROVED** [review]
LGTM, thank you for contribution :)

**Comment by JuankleChan** [comment]
can this commit merger into v0.3 & v0.4 branch ?

**Comment by arugal** [comment]
> can this commit merger into v0.3 & v0.4 branch ?

`v0.3.0` and `v0.4.0` is tag.

**Comment by JuankleChan** [comment]
> > can this commit merger into v0.3 & v0.4 branch ?
> 
> `v0.3.0` and `v0.4.0` is tag.

what shoud i do if i want tracer sampler but the oap version must be v6.4.0? is it possible to new a branch for adapting the old grcp protocol?

**Pull request merged**